### ISSUE TITLE
Inconsistent error when running pyright on an entire project vs. the offending file

### ISSUE THREAD

**Issue by twoertwein** [microsoft/pyright#2315]
**Describe the bug**
Pyright complains about a "TypeVar bound type cannot be generic" when checking the entire pandas project but when checking only the offending file, pyright seems to be happy. In both cases, many/most type checking rules are disabled.

The output of running [pyright on the pandas code base](https://github.com/pandas-dev/pandas/runs/3656056815) when [trying to add pyright to pandas's pre-commit](https://github.com/pandas-dev/pandas/pull/43672): 
[pandas/core/computation/expr.py:268](https://github.com/pandas-dev/pandas/blob/ae049aea7b83a63dddad42fae486e44e827025be/pandas/core/computation/expr.py#L268):26 - error: TypeVar bound type cannot be generic

**To Reproduce**
I hoped that the following example would reproduce the issue but it doesn't seem to trigger it:

```py
from __future__ import annotations

from typing import Callable, TypeVar

_T = TypeVar("_T", bound="Test")

def disallow() -> Callable[[type[_T]], type[_T]]:
    ...

@disallow()
class Test:
    ...
```

**Comment by erictraut** [comment]
The problem is due to a circular type dependency. The file `pandas/core/computation/expr.py` defines a TypeVar called `_T` that is bound to the type `BaseExprVisitor`. This TypeVar is used in the declaration of the `disallow` decorator function. But this same decorator function is used to decorate `BaseExprVisitor`, so evaluating its type requires the decorator to be evaluated.

The correct fix is to break this cyclical dependency. You can do this by removing the `disallow` decorator call on `BaseExprVisitor` or removing the bound on the TypeVar `_T`.

The cause of the problem would be more evident if you hadn't disabled so many classes of errors, including the `reportGeneralTypeIssues` diagnostic rule. This rule covers the most basic of type violation rules, so I recommend that you don't disable it.

**Comment by twoertwein** [comment]
Thank you very much for debugging this!

> The cause of the problem would be more evident if you hadn't disabled so many classes of errors, including the `reportGeneralTypeIssues` diagnostic rule. This rule covers the most basic of type violation rules, so I recommend that you don't disable it.

Do you have recommendations for pandas how to make its code slowly more pyright compatible? My aim with the PR is to enable pyright with a very very small ruleset (`reportGeneralTypeIssues` would generate too many errors). And then make PRs that enable certain rules one by one (such as `reportInvalidTypeVarUse`). Initially, I see pyright's role in pandas as preventing type regressions that are currently not captured by pandas's mypy setup.

I saw your comment with the per-file annotions https://github.com/microsoft/pyright/issues/1443#issuecomment-770423431. I think pandas isn't adding new files that often, but of course pyright could be run only on selected files.
